### PR TITLE
Cknieling/annotation attribute UI

### PR DESCRIPTION
**Pull request by chr11115**
- added support for restricted/unrestricted range annotations
- addes support for file extension filtering

### CODE REVIEW

File: src/palladio/AnnotationParsing.cpp [hunk 1]
```diff
@@ -270,10 +268,21 @@ AttributeAnnotationInfo getAttributeAnnotationInfo(const std::wstring& attribute
 					annotation = currAnnotation;
 					attributeTrait = currAttributeTrait;
 				}
-			}	
+			}
 			else {
-				annotation = currAnnotation;
-				attributeTrait = currAttributeTrait;
+				//keep annotation of range and trait of percent/angle
+				const bool keepAnnotation = (((currAttributeTrait == AttributeTrait::PERCENT) ||
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.cpp:274]: the handling of percent/angle/range is hard to read and somewhat convoluted... how about just parsing the annotation traits into an enum set and then handling this when creating the spare parms?

**Review by mistafunk — COMMENTED** [review]
* see inline comments
* semi-related: in AnnotationParsing.cpp:155, 160, 165, use emplace_back and remove the .c_str() to avoid extra string construction
* semi-related optimization: getAttributeAnnotationInfo and its call site have quadratic cost when looping over the attributes -> consider parsing the annotation info once and storing a map

File: src/palladio/NodeSpareParameter.cpp [hunk 2]
```diff
@@ -40,6 +40,36 @@ std::string getUniqueIdFromFolderVec(const FolderVec& parentFolders) {
...
+			continue;
+		if (extensionString.empty())
+			extensionString += cleanedFileExtension;
+		else
+			extensionString += " " + cleanedFileExtension;
+	}
+	if (extensionString.empty())
+		return FILE_ANY;
+	return extensionString;
+}
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp]: done

File: src/palladio/SOPAssign.cpp [hunk 3]
```diff
@@ -604,27 +604,31 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
...
+						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, rangeAnnotation.min,
+						                                 rangeAnnotation.max, rangeAnnotation.restricted, parentFolders,
+						                                 description);
 						break;
 					}
 					case AnnotationParsing::AttributeTrait::ANGLE: {
-						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, 0, 360, parentFolders,
-						                                 description);
+						AnnotationParsing::RangeAnnotation rangeAnnotation =
+						        AnnotationParsing::parseRangeAnnotation(annotationInfo.mAnnotation, 0, 360);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: fixed

File: src/palladio/SOPAssign.cpp [hunk 4]
```diff
@@ -604,27 +604,31 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
...
+						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, rangeAnnotation.min,
+						                                 rangeAnnotation.max, false, parentFolders, description);
 						break;
 					}
 					case AnnotationParsing::AttributeTrait::PERCENT: {
-						// diplay % values with a 0-1 range for now (avoid scaling by 100)
-						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, 0, 1, parentFolders,
-						                                 description);
+						AnnotationParsing::RangeAnnotation rangeAnnotation =
+						        AnnotationParsing::parseRangeAnnotation(annotationInfo.mAnnotation, 0, 1);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: fixed

File: src/palladio/SOPAssign.cpp [hunk 5]
```diff
@@ -604,27 +604,31 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
...
+						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, rangeAnnotation.min,
+						                                 rangeAnnotation.max, false, parentFolders, description);
 						break;
 					}
 					default: {
-						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, 0, 10, parentFolders,
-						                                 description);
+						NodeSpareParameter::addFloatParm(this, attrId, attrName, defaultValue, 0, 10, true,
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: fixed

File: src/palladio/AnnotationParsing.h [hunk 6]
```diff
@@ -54,7 +58,9 @@ ColorAnnotation parseColor(const std::wstring colorString);
 
 std::wstring getColorString(const std::array<float, 3>& rgb);
 
-RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation);
+RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation,
+                                     double fallbackMin = std::numeric_limits<double>::quiet_NaN(),
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.h:62]: nolonger necessary due to the decoupling of annotations

File: src/palladio/NodeSpareParameter.cpp [hunk 7]
```diff
@@ -138,9 +138,10 @@ void addParm(OP_Node* node, PRM_Type parmType, const std::wstring& id, const std
 }
 
 void addFloatParm(OP_Node* node, const std::wstring& id, const std::wstring& name, double defaultVal, double min,
-                  double max, const FolderVec& parentFolders, const std::wstring& description) {
+                  double max, bool restricted, const FolderVec& parentFolders, const std::wstring& description) {
+	PRM_RangeFlag restrictedFlag = restricted ? PRM_RANGE_RESTRICTED : PRM_RANGE_UI;
 	PRM_Range range =
-	        (!std::isnan(min) && !std::isnan(max)) ? PRM_Range(PRM_RANGE_UI, min, PRM_RANGE_UI, max) : PRM_Range();
+	        (!std::isnan(min) && !std::isnan(max)) ? PRM_Range(restrictedFlag, min, restrictedFlag, max) : PRM_Range();
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/NodeSpareParameter.cpp:115]: nolonger necessary due to the decoupling of annotations

File: src/palladio/AnnotationParsing.cpp [hunk 8]
```diff
@@ -97,38 +97,40 @@ wchar_t toHex(int i) {
 } // namespace
 
 namespace AnnotationParsing {
-RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation) {
-	auto minMax = std::make_pair(std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN());
+RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation, double fallbackMin, double fallbackMax) {
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp]: Tbh I don't really see the benefit of encapsulating minMax in a pair, if we already have a struct field. I seems just more cumbersome to access min via struct.minMax.first/ create the struct via {make_pair(min, max), restricted}.
I have created a simple commit with the changes, but if you agree I can just roll back that commit..

File: src/palladio/AnnotationParsing.cpp (hunk 1 above)
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:274]: Now using a different approach with a map from the type enum to a std::variant that contains the annotation information.

File: src/palladio/NodeSpareParameter.h [hunk 9]
```diff
@@ -23,15 +23,17 @@
...
+             const std::wstring& description = {});
 void addFloatParm(OP_Node* node, const std::wstring& id, const std::wstring& name, double defaultVal,
                   double min = std::numeric_limits<double>::quiet_NaN(),
-                  double max = std::numeric_limits<double>::quiet_NaN(), const FolderVec& parentFolders = {},
-                  const std::wstring& description = {});
+                  double max = std::numeric_limits<double>::quiet_NaN(), bool restricted = true, bool isPercent = false,
+                  const FolderVec& parentFolders = {}, const std::wstring& description = {});
 void addColorParm(OP_Node* node, const std::wstring& id, const std::wstring& name, std::array<double, 3> defaultVal,
                   const FolderVec& parentFolders, const std::wstring& description = {});
 void addIntParm(OP_Node* node, const std::wstring& id, const std::wstring& name, int defaultVal,
```
Comment (mistafunk) [inline src/palladio/NodeSpareParameter.h:39]: addIntParm is not used, should we remove it?

File: src/palladio/AnnotationParsing.h [hunk 10]
```diff
@@ -32,23 +35,14 @@ struct EnumAnnotation {
 	std::vector<std::wstring> mOptions;
 };
 
-using RangeAnnotation = std::pair<double, double>;
+struct RangeAnnotation {
+	std::pair<double, double> minMax;
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.h:39]: please default initialize it to NaN, NaN

File: src/palladio/AnnotationParsing.cpp [hunk 11]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
 
 namespace AnnotationParsing {
 RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation) {
-	auto minMax = std::make_pair(std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN());
+	double min = std::numeric_limits<double>::quiet_NaN();
+	double max = std::numeric_limits<double>::quiet_NaN();
+	std::pair<double, double> minMax;
+	bool restricted = false;
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.cpp:104]: Consider directly introducing a default initialized RangeAnnotation object

File: src/palladio/AnnotationParsing.h [hunk 12]
```diff
@@ -32,23 +35,14 @@ struct EnumAnnotation {
...
-
-	AttributeAnnotationInfo(const prt::Annotation& annotation, AttributeTrait attributeTrait, const std::wstring& description)
-	    : mAnnotation(annotation), mAttributeTrait(attributeTrait), mDescription(description) {}
-	AttributeAnnotationInfo() = delete;
-	AttributeAnnotationInfo(const AttributeAnnotationInfo&) = delete;
-	AttributeAnnotationInfo(AttributeAnnotationInfo&&) = delete;
-	AttributeAnnotationInfo& operator=(const AttributeAnnotationInfo&) = delete;
-	AttributeAnnotationInfo& operator=(AttributeAnnotationInfo&&) = delete;
-};
+using AnnotationVariant = std::variant<std::monostate, EnumAnnotation, RangeAnnotation, FileAnnotation, std::wstring>;
```
Comment (mistafunk) [inline src/palladio/AnnotationParsing.h:45]: This name is not very intuitive to me. How about `AnnotationTraitParameter` or `AnnotationParameter`?

Similarly, `AttributeTrait` above is too generic to me, how about `AnnotationTrait`?

File: src/palladio/SOPAssign.cpp [hunk 13]
```diff
@@ -554,8 +566,16 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
 		const std::wstring attrName = ra.niceName;
 		const std::wstring attrId = toUTF16FromOSNarrow(NameConversion::toPrimAttr(ra.fqName).toStdString());
 
-		const auto annotationInfo = AnnotationParsing::getAttributeAnnotationInfo(ra.fqName, ruleFileInfo);
-		const std::wstring description = annotationInfo.mDescription;
+		const auto& annotationsIt = attributeAnnotations.find(ra.fqName);
+		if (annotationsIt == attributeAnnotations.end())
+			continue;
+		auto& traitAnnotationMap = annotationsIt->second;
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:572]: const

File: src/palladio/SOPAssign.cpp [hunk 14]
```diff
@@ -554,8 +566,16 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
...
+
+		const auto& enumIt = traitAnnotationMap.find(AnnotationParsing::AttributeTrait::ENUM);
+		const auto& descriptionIt = traitAnnotationMap.find(AnnotationParsing::AttributeTrait::DESCRIPTION);
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:575]: too many `auto`s for my taste, maybe introducing type aliases for the `attributeAnnotations` map type?

**Review by mistafunk — CHANGES_REQUESTED** [review]
I feel the annotation parsing code is not yet clean enough, see my comments.

Also, the `buildGUI` function is 150 lines long and hard to read. Consider introducing `tryHandleAngle` etc functions....

File: src/palladio/NodeSpareParameter.h (hunk 9 above)
Comment (chr11115) [reply src/palladio/NodeSpareParameter.h:39]: I will remove it. Initially it was not clear to me if we needed it.

File: src/palladio/AnnotationParsing.cpp (hunk 11 above)
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:104]: using default initialized RangeAnnotation object.

File: src/palladio/AnnotationParsing.cpp [hunk 15]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
 
 namespace AnnotationParsing {
 RangeAnnotation parseRangeAnnotation(const prt::Annotation& annotation) {
-	auto minMax = std::make_pair(std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN());
+	double min = std::numeric_limits<double>::quiet_NaN();
+	double max = std::numeric_limits<double>::quiet_NaN();
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:102]: done

File: src/palladio/AnnotationParsing.cpp [hunk 16]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
...
+	std::pair<double, double> minMax;
+	bool restricted = false;
 
 	const wchar_t* anName = annotation.getName();
 
 	if (std::wcscmp(anName, ANNOT_RANGE) == 0) {
 		const RangeType annotationRangeType = GetRangeType(annotation);
 		if (annotationRangeType != RangeType::RANGE)
-			return minMax;
+			return {std::make_pair(min, max), restricted};
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:111]: done

File: src/palladio/AnnotationParsing.cpp [hunk 17]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
...
-			else if (std::wcscmp(key, MAX_KEY) == 0) {
-				minMax.second = arg->getFloat();
-			}
+			if (std::wcscmp(key, MIN_KEY) == 0)
+				min = arg->getFloat();
+			else if (std::wcscmp(key, MAX_KEY) == 0)
+				max = arg->getFloat();
+			else if (std::wcscmp(key, RESTRICTED_KEY) == 0)
+				restricted = arg->getBool();
 		}
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:124]: done

File: src/palladio/AnnotationParsing.cpp [hunk 18]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
...
 
 		// parse old style range
-		if ((std::isnan(minMax.first) || std::isnan(minMax.second)) && (numArgs == 2)) {
-			minMax.first = annotation.getArgument(0)->getFloat();
-			minMax.second = annotation.getArgument(1)->getFloat();
-		}
+		if ((std::isnan(min) || std::isnan(max)) && (numArgs == 2))
+			minMax = std::make_pair(annotation.getArgument(0)->getFloat(), annotation.getArgument(1)->getFloat());
+		else
+			minMax = std::make_pair(min, max);
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:130]: done

File: src/palladio/AnnotationParsing.cpp [hunk 19]
```diff
@@ -98,37 +98,40 @@ wchar_t toHex(int i) {
...
 
-		return minMax;
+		return {minMax, restricted};
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:132]: removed

File: src/palladio/AnnotationParsing.h (hunk 12 above)
Comment (chr11115) [reply src/palladio/AnnotationParsing.h:45]: renamed to AnnotationTraitParameter

File: src/palladio/SOPAssign.cpp (hunk 13 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:572]: changed

File: src/palladio/AnnotationParsing.cpp [hunk 20]
```diff
@@ -239,48 +258,31 @@ AttributeTrait detectAttributeTrait(const prt::Annotation& annotation) {
 	return AttributeTrait::NONE;
 }
 
-AttributeAnnotationInfo getAttributeAnnotationInfo(const std::wstring& attributeName, const RuleFileInfoUPtr& info) {
+std::map<std::wstring, std::map<AttributeTrait, AnnotationVariant>>
+getAttributeAnnotations(const RuleFileInfoUPtr& info) {
+	std::map<std::wstring, std::map<AttributeTrait, AnnotationVariant>> annotationMap;
+
 	const prt::Annotation* annotation = nullptr;
 	AttributeTrait attributeTrait = AttributeTrait::NONE;
```
[the remark this answers is not in the document]
Comment (chr11115) [reply src/palladio/AnnotationParsing.cpp:266]: removed.

File: src/palladio/SOPAssign.cpp (hunk 14 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:575]: These are all iterators returned by "map.find()"
I replaced the map type in the header and cpp of AnnotationParsing aswell as the one time it's used in SOPAssign.

File: src/palladio/SOPAssign.cpp [hunk 21]
```diff
@@ -398,6 +400,119 @@ bool evaluateDefaultRuleAttributes(SOPAssign* node, const GU_Detail* detail, Sha
...
+double getDefaultBool(const SOPAssign::CGAAttributeValueType& defaultValue) {
+	if (std::holds_alternative<bool>(defaultValue))
+		return std::get<bool>(defaultValue);
+	return false;
+}
+
+double getDefaultFloat(const SOPAssign::CGAAttributeValueType& defaultValue) {
+	if (std::holds_alternative<double>(defaultValue))
+		return std::get<double>(defaultValue);
+	return 0;
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:412]: better write 0.0 as well

File: src/palladio/SOPAssign.cpp [hunk 22]
```diff
@@ -398,6 +400,119 @@ bool evaluateDefaultRuleAttributes(SOPAssign* node, const GU_Detail* detail, Sha
 	return true;
 }
 
+double getDefaultBool(const SOPAssign::CGAAttributeValueType& defaultValue) {
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:403]: wrong return type? should be bool?

File: src/palladio/SOPAssign.cpp [hunk 23]
```diff
@@ -398,6 +400,119 @@ bool evaluateDefaultRuleAttributes(SOPAssign* node, const GU_Detail* detail, Sha
...
+std::wstring getDefaultString(const SOPAssign::CGAAttributeValueType& defaultValue) {
+	if (std::holds_alternative<std::wstring>(defaultValue))
+		return std::get<std::wstring>(defaultValue);
+	return {};
+}
+
+std::wstring getDescription(const AnnotationParsing::TraitParameterMap& traitParmMap) {
+	const auto& descriptionIt = traitParmMap.find(AnnotationParsing::AttributeTrait::DESCRIPTION);
+	if (descriptionIt != traitParmMap.end())
+		return std::get<std::wstring>(descriptionIt->second);
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:424]: this get could (theoretically) throw as we do not check for the variant type

there are some more cases below, maybe wrap in try/catch or use std::get_if?

File: src/palladio/SOPAssign.cpp [hunk 24]
```diff
@@ -554,124 +680,69 @@ void SOPAssign::buildUI(GU_Detail* detail, ShapeData& shapeData, const ShapeConv
...
+		const std::wstring& description = getDescription(traitParmMap);
 
 		FolderVec parentFolders;
 		parentFolders.push_back(RULE_ATTRIBUTES_FOLDER_NAME);
 		parentFolders.push_back(ra.ruleFile);
 		parentFolders.insert(parentFolders.end(), ra.groups.begin(), ra.groups.end());
 
-		const auto defaultValIt = mDefaultCGAAttributes.find(ra.fqName);
+		const auto& defaultValIt = mDefaultCGAAttributes.find(ra.fqName);
 		const bool foundDefaultValue = (defaultValIt != mDefaultCGAAttributes.end());
```
Comment (mistafunk) [inline src/palladio/SOPAssign.cpp:696]: unused variable

**Review by mistafunk — CHANGES_REQUESTED** [review]
Almost there...

File: src/palladio/SOPAssign.cpp (hunk 23 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:424]: using std::holds_alternative in the if above, like in the other examples

File: src/palladio/SOPAssign.cpp (hunk 22 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:403]: fixed

File: src/palladio/SOPAssign.cpp (hunk 21 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp:412]: done.

File: src/palladio/SOPAssign.cpp (hunk 24 above)
Comment (chr11115) [reply src/palladio/SOPAssign.cpp]: removed

**Pull request merged**